pkg/autoscaler/autoscaler/calculate/calculate.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package calculate
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	promClient "github.com/prometheus/client_golang/api"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+const (
+	TikvSumCpuMetricsPattern     = `sum(tikv_thread_cpu_seconds_total{cluster="%s"}) by (instance)`
+	TidbSumCpuMetricsPattern     = `sum(process_cpu_seconds_total{cluster="%s",job="tidb"}) by (instance)`
+	InvalidTacMetricConfigureMsg = "tac[%s/%s] metric configuration invalid"
+	CpuSumMetricsErrorMsg        = "tac[%s/%s] cpu sum metrics error, can't calculate the past %s cpu metrics, may caused by prometheus restart while data persistence not enabled"
+	queryPath                    = "/api/v1/query"
+
+	float64EqualityThreshold = 1e-9
+)
+
+func queryMetricsFromPrometheus(tac *v1alpha1.TidbClusterAutoScaler,
+	query string, client promClient.Client, timestamp int64, resp *Response) error {
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s%s", *tac.Spec.MetricsUrl, queryPath), nil)
+	if err != nil {
+		return err
+	}
+	q := req.URL.Query()
+	q.Add("query", query)
+	q.Add("time", fmt.Sprintf("%d", timestamp))
```

Comment:
With this query, the end time is `time` and what's the start time? Do we need to specify a range?

Reply:
No, the metrics is to query the total counts from Prometheus initialization to the given timestamp. To query the D-value between the start time and end time is to query it twice with different timestamp.